[Feature:Forum] Live Chat Room

hansongu123 opened the pull request:
### What is the new feature and why is iit important?

Live text chat/discussion between instructor, (possible TAs), and students was useful during pandemic 100% remote live video lectures. This feature was integrated with Webex, Zoom, Google Meetings, etc.

When we return to the fully in-person classroom, having this same digital chat can be beneficial:

In a large classroom, it can be hard to hear the question coming from a person in the back of the room. Having the question typed out can improve accuracy of communication.
Shy students may wish to stay (mostly) anonymous, but still need to ask a question or want to answer the instructor's question to the class.
A student may want a specific detail of lecture or an announcement repeated or clarified, and another student in the course can do that.

Closes #7207.
Future PRs will refine and improve this feature.

### What is the new behavior?

Steps to enable the feature:

run submitty_install
In instructor's UI:

go to course settings
<img src="https://github.com/Submitty/Submitty/assets/123511202/41e80d16-01f3-4208-96e6-0231350bb62d" width=70% height=70%>

scroll down to toggle on live lecture chat:
<img src="https://github.com/Submitty/Submitty/assets/123511202/6d291113-cbf6-4d9b-8566-b170f5691e3e" width=70% height=70%>

Now a "Live Lecture Chat" option should be available on the left side menu.

Toggle start session to activate a chat room, then the chat room will be available for the students to join.
![image](https://github.com/user-attachments/assets/b8a1f904-6c52-4dff-a611-e2562cc5a2a5)

------------------------------------------------------------------------------------------------------------------------------------
Screenshots
![image](https://github.com/user-attachments/assets/61d8fbcf-1dbb-4bc9-8b25-4b929ff9813e)

![image](https://github.com/Submitty/Submitty/assets/123511202/189ef1bb-6d2d-46e0-a794-971f007f4745)

![454499560-3a65158d-8a4a-4c6c-bacb-7904e8bd797c](https://github.com/user-attachments/assets/55581d87-f022-46cb-8a8b-709f2f1f42e2)


Feature includes:
Anonymous join - enable/disable in instructor's UI upon creating an editing a chatroom

Notes:
The dynamic display of messages is currently through the use of WebSockets.

cjreed121 commented on migration/migrator/migrations/course/20240209021641_lecture_chat.py:
```diff
@@ -0,0 +1,48 @@
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("DROP TABLE IF EXISTS chatrooms")
+    database.execute("DROP TABLE IF EXISTS chatroom_messages")
```
We shouldn't drop the tables so we can prevent any data loss.

hansongu123 replied:
I'm currently still making adjustments to the tables, will remove these two lines after the PR is converted to open

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,187 @@
+    $em = $this->core->getCourseEntityManager();
+    $chatroom = $em->getRepository(Chatroom::class)->find($chatroom_id);
+
+    if (!$chatroom) {
```
Direct comparison on null would be better

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,187 @@
+    }
+
+/**
+ * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/edit", name="edit_chatroom", methods={"POST"})
```
We don't ever use Route names so you could remove those from the docblock. Not necessary to remove them but I'm just pointing out that they don't do anything for us. So I'll leave it up to you.

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,187 @@
+        $chatroom->setDescription($_POST['description']);
+    }
+
+    $em->persist($chatroom);
```
You don't need to persist again. You only need to persist when first creating an object.

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,187 @@
+            return JsonResponse::getFailResponse('Invalid Chatroom ID');
+        }
+
+        foreach ($chatroom->getMessages() as $message) {
+            $em->remove($message);
+        }
```
Will this function work without this because of the cascade you have in the entity?

hansongu123 replied:
I'll remove either this or cascade

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,187 @@
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/messages", name="fetch_chatroom_messages", methods={"GET"})
+     */
+    public function fetchMessages(int $chatroom_id): JsonResponse {
+        $em = $this->core->getCourseEntityManager();
+        $messages = $em->getRepository(Message::class)->findBy(['chatroom' => $chatroom_id], ['timestamp' => 'ASC']);
+
+        $formattedMessages = array_map(function ($message) {
+            return [
+                'id' => $message->getId(),
+                'content' => $message->getContent(),
+                'timestamp' => $message->getTimestamp()->format('Y-m-d H:i:s'),
+                'user_id' => $message->getUserId(),
+                'display_name' => $message->getDisplayName(),
+                'role' => $message->getRole()
+            ];
+        }, $messages);
+
+        return JsonResponse::getSuccessResponse($formattedMessages);
+    }
```
Will there be a restriction for only messages from that day?

hansongu123 replied:
Good question. Currently there's no restriction to a student's access to the history of messages. We could either set this restriction to either day-based, session-based, or decide something else.

cjreed121 commented on site/app/entities/chat/Message.php:
```diff
@@ -0,0 +1,95 @@
+    #[ORM\GeneratedValue]
+    private int $id;
+
+    #[ORM\ManyToOne(targetEntity: "Chatroom")]
```
For target entity you can use the ::class syntax to specifically refer to that class. It is better for linting purposes and would be helpful if we ever changed the directory structure.

cjreed121 commented on site/app/views/ChatroomView.php:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public function showChatPageInstructor(array $chatrooms) {
+        return $this->core->getOutput()->renderTwigTemplate("chat/ChatPageIns.twig", [
+            'csrf_token' => $this->core->getCsrfToken(),
+            'base_url' => $this->core->buildCourseUrl() . '/chat',
+            'semester' => $this->core->getConfig()->getTerm(),
+            'course' => $this->core->getConfig()->getCourse(),
+            'chatrooms' => $chatrooms
+        ]);
+    }
+
+    public function showChatPageStudent(array $chatrooms) {
+        return $this->core->getOutput()->renderTwigTemplate("chat/ChatPageStu.twig", [
+            'csrf_token' => $this->core->getCsrfToken(),
+            'base_url' => $this->core->buildCourseUrl() . '/chat',
+            'semester' => $this->core->getConfig()->getTerm(),
+            'course' => $this->core->getConfig()->getCourse(),
+            'chatrooms' => $chatrooms
+        ]);
```
These are very similar (and the contents in the twig files are very similar too). Could we combine them and just do some conditional rendering in the twig file?

hansongu123 replied:
I'll try to reduce redundancy. I wanted more clarity at this point because student's and instructor's UI could be more different.

williamjallen replied:
Although it may be tempting to have separate student and instructor UIs, we learned from the polls page that that is a bad idea.  Even though it adds additional complexity, it is important to have all of the information for a given page in one place.

cjreed121 commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,176 @@
+    if (role === "instructor") {
+        message.classList.add('admin-message')
+    }
+    message.innerHTML = `
+        <div class="message-header">
+            <span class="sender-name">${display_name}</span>
+            <span class="timestamp">${timestamp}</span>
+        </div>
+        <div class="message-content">
+            ${content}
+        </div>
+    `;
```
This is dangerous to do because you are using .innerHTML. This could be subject to an XSS attack. Any variable that contains data coming from the user should be sanitized. Easiest way to do this is to create any needed elements and then insert user data with .innerText or .text() if using jQuery.

cjreed121 reviewed (changes requested):
Code looks pretty good overall but I do have some requested changes.

There is also another security concern where any user can send messages over the websocket server to all other users without it being logged. They can also pretend to be other users too. In order to address this, the websocket messages must be sent from the PHP side and we must not allow clients to send these messages. The latter part might need to be done in another PR. You should be able to move the websocket send message into PHP easily by referring to either office hour or forum code.

williamjallen commented on migration/migrator/migrations/course/20240209021641_lecture_chat.py:
```diff
@@ -0,0 +1,48 @@
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("CREATE TABLE IF NOT EXISTS chatrooms(id SERIAL PRIMARY KEY, host_id character varying NOT NULL, host_name character varying, title text NOT NULL, description text, is_active BOOLEAN DEFAULT false NOT NULL, allow_anon BOOLEAN DEFAULT true NOT NULL)")
+    database.execute("CREATE TABLE IF NOT EXISTS chatroom_messages(id SERIAL PRIMARY KEY, chatroom_id integer NOT NULL, userId character varying NOT NULL, display_name character varying, role character varying, content text NOT NULL, timestamp timestamp(0) with time zone NOT NULL)")
```
Can you wrap these lines to make them more readable on most displays?

williamjallen commented on migration/migrator/migrations/course/20240209021641_lecture_chat.py:
```diff
@@ -0,0 +1,48 @@
+    database.execute("CREATE TABLE IF NOT EXISTS chatroom_messages(id SERIAL PRIMARY KEY, chatroom_id integer NOT NULL, userId character varying NOT NULL, display_name character varying, role character varying, content text NOT NULL, timestamp timestamp(0) with time zone NOT NULL)")
+
+    course_dir = Path(config.submitty['submitty_data_dir'], 'courses', semester, course)
+    # add boolean to course config
+    config_file = Path(course_dir, 'config', 'config.json')
+    if config_file.is_file():
+        with open(config_file, 'r') as in_file:
+            j = json.load(in_file)
+            j['course_details']['chat_enabled'] = False
+
+        with open(config_file, 'w') as out_file:
+            json.dump(j, out_file, indent=4)
```
Instead of editing all exiting configs, why can't the default value be false instead?  The field could then be optional.

williamjallen commented on migration/migrator/migrations/course/20240209021641_lecture_chat.py:
```diff
@@ -0,0 +1,48 @@
+    :type course: str
+    """
+    database.execute("CREATE TABLE IF NOT EXISTS chatrooms(id SERIAL PRIMARY KEY, host_id character varying NOT NULL, host_name character varying, title text NOT NULL, description text, is_active BOOLEAN DEFAULT false NOT NULL, allow_anon BOOLEAN DEFAULT true NOT NULL)")
+    database.execute("CREATE TABLE IF NOT EXISTS chatroom_messages(id SERIAL PRIMARY KEY, chatroom_id integer NOT NULL, userId character varying NOT NULL, display_name character varying, role character varying, content text NOT NULL, timestamp timestamp(0) with time zone NOT NULL)")
```
These need foreign keys to be set up wherever possible.  Chatroom messages should have foreign-key relationships with chatrooms and users at a minimum.

As an aside: what is the purpose of the `host_id` and `host_name` columns?

hansongu123 replied:
These columns currently don't have other purposes except for showing the name of the host on students' UI. I intended. to include them to prepare for if we plan to have different hosts hosting different chat rooms in the future.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+
+class ChatroomController extends AbstractController {
+
+    public function __construct(Core $core) {
+        parent::__construct($core);
+    }
```
This isn't necessary.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/newChatroom", methods={"POST"})
```
`.../chat/newChatroom` seems a bit redundant.  What about `.../chat/new`?

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+    public function addChatroom(): RedirectResponse {
+        $em = $this->core->getCourseEntityManager();
+
+        $chatroom = new Chatroom();
+        $chatroom->setTitle($_POST['title']);
+        $chatroom->setDescription($_POST['description']);
+        $chatroom->setHostId($this->core->getUser()->getId());
+        $chatroom->setHostName($this->core->getUser()->getDisplayFullName());
```
Would this be a good opportunity to use a constructor?  See the `Poll` type for an example.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+        }
+
+        $repo = $this->core->getCourseEntityManager()->getRepository(Chatroom::class);
+        $chatroom = $repo->find($chatroom_id);
```
What if no chat room was found?

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+        );
+     }
+
+     /**
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/anonymous", methods={"GET"})
+     * @param string $chatroom_id
+     * @return RedirectResponse|WebResponse
+     */
+     public function getChatroomAnon(string $chatroom_id): WebResponse|RedirectResponse {
```
What's the purpose of this route?  It appears to be essentially identical to the route without the `/anonymous`

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+     }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/deleteChatroom", methods={"POST"})
```
What about `.../chat/delete`?

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+     * @Route("/courses/{_semester}/{_course}/chat/deleteChatroom", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function deleteChatroom(): JsonResponse {
```
Should we have the ability to delete chatrooms?  Perhaps we should keep them around but add a boolean "deleted" column in case there's ever a desire to dig up the chatroom in the future.

hansongu123 replied:
I thought of this before. I added a column in the database table to allow a chat room to be "session" based. Toggle on a chat room will allow it to be active and available to the students, and stopping the session will take away students' ability to type into the chat and/or entering the chat room. So instead of a "deleted" column, a chat room would just be "inactive", and the host and activate it again easily if they wish to dig it up or reuse it.

On the other hand if a chat room is deleted it is deleted, so there's no confusion.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/edit", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function editChatroom(int $chatroom_id): RedirectResponse {
```
Will the `int` type hint work here?  You used `string` earlier in this file, and I seem to recall that the type is always `string` for query parameters in Submitty.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+     * @Route("/courses/{_semester}/{_course}/chat/{chatroom_id}/toggleOnOff", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+     public function toggleChatroomOnOf(int $chatroom_id): RedirectResponse {
```
```suggestion
     public function toggleChatroomOnOff(int $chatroom_id): RedirectResponse {
```

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+        $em = $this->core->getCourseEntityManager();
+        $messages = $em->getRepository(Message::class)->findBy(['chatroom' => $chatroom_id], ['timestamp' => 'ASC']);
+
+        $formattedMessages = array_map(function ($message) {
+            return [
+                'id' => $message->getId(),
+                'content' => $message->getContent(),
+                'timestamp' => $message->getTimestamp()->format('Y-m-d H:i:s'),
+                'user_id' => $message->getUserId(),
+                'display_name' => $message->getDisplayName(),
+                'role' => $message->getRole()
+            ];
+        }, $messages);
```
I believe using `toArray()` on the Doctrine collection will have the same effect.

williamjallen commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,234 @@
+     */
+    public function addMessage(int $chatroom_id): JsonResponse {
+        $em = $this->core->getCourseEntityManager();
+        $content = $_POST['content'] ?? '';
```
Should we allow people to send blank messages?

williamjallen replied:
It would probably be good to add constraints like this to the database table as well.

williamjallen commented on site/app/models/Chatroom.php:
Is there a reason why there's an empty file here?

hansongu123 replied:
nope

williamjallen commented on site/app/repositories/chat/ChatroomRepository.php:
```diff
@@ -0,0 +1,35 @@
+
+class ChatroomRepository extends EntityRepository {
+
+    public function findAllChatroomsByHostId(string $hostId) {
+        return $this->createQueryBuilder('c')
+                    ->where('c.host_id = :hostId')
+                    ->setParameter('hostId', $hostId)
+                    ->getQuery()
+                    ->getResult();
+    }
+
+    public function findAllActiveChatrooms() {
+        return $this->createQueryBuilder('c')
+                    ->where('c.isActive = :isActive')
+                    ->setParameter('isActive', true)
+                    ->getQuery()
+                    ->getResult();
+    }
+
+    public function findAllInactiveChatrooms() {
+        return $this->createQueryBuilder('c')
+                    ->where('c.isActive = :isActive')
+                    ->setParameter('isActive', false)
+                    ->getQuery()
+                    ->getResult();
+    }
```
I believe it's recommended to use DQL instead of the query builder for performance.

williamjallen reviewed (changes requested):
I haven't finished reviewing the entire PR yet, but here's my first round of comments.  In general, things look very good, but we need to make sure everything is robust since this will be exposed to hundreds of students at a time.  This also needs to have some tests added.  For now, some basic Cypress tests would be good at least, if only to ensure that there are no repeated queries occurring behind the scenes.  Any sort of repeated queries would be detrimental to the success of this feature.

RayReyRam reviewed (changes requested):
I switched to this branch and ran a sumbitty_install as instructor went to sample and have this error preventing me to enter 
<img width="1078" alt="error" src="https://github.com/Submitty/Submitty/assets/148298134/f3537603-6d55-4691-8c88-87cfed1169ce">

hansongu123 commented:
> I switched to this branch and ran a sumbitty_install as instructor went to sample and have this error preventing me to enter 

@RayReyRam I updated the migration and the dump file, can you pull from this branch and try again? If that doesn't work, try (in vagrant ssh) calling `migrator -e course rollback` and` migrator -e course migrate` to manually apply the migration. If none of this works, try re-creating the sample courses: 
`sudo bash /usr/local/submitty/GIT_CHECKOUT/Submitty/.setup/bin/recreate_sample_courses.sh --no_submissions` in vagrant ssh.

@williamjallen Is this caused by the removal of the script in the migration file that edits all exiting configs?

RayReyRam reviewed (changes requested):
Manually migrated worked.
I believe that live chat feature should be on by default, as I had trouble finding it and stumbled upon it luckily in settings. 

I also think there should be a reply feature, maybe a student knows the answer to another student's question, or a ta directly wants to answer and could reply to it. 

Just a thought but maybe have some sort of tag feature, like if it's a question a tag could be placed making easier for the instructor/ta to see. 

I found an error that when the instructor ends the session of the live chat, the student is still able to message in it as long they stay on it. Or just go back a page. They should be kicked out the live chat, unless that is intended.

Also don't remember if this was talked about but allowing a student to view a previously closed session. This could be a seeing toggled by the instructor, but i feel like it would be useful if a student can go back into the chat especially if questions were answered in the session.

Other than that the live chat seems to work I'm able to message through 2 accounts without any problem.

hansongu123 commented:
@RayReyRam Thanks for the review, I find it really helpful! I added a step in the PR description for how to enable it as a sidebar item. I haven't added the mechanism to disable students to entering messages upon terminating a session, or limiting students' access to the entire chat history just yet. I would like to hear back from the reviewers and decide how to design these mechanisms.

JoJo-ESC reviewed (commented):
Overall I love this idea and I think it was executed exceptionally well. There are a couple things that I would like to see changed before merged.

If live chat won't be enabled automatically, please, if possible, move the live lecture on option to be right below the forum option. I think grouping things that deal with similar ideas imporves clarity. Additionally Im not sure if this idea was discussed already but I really think it might be worth it to consider implementing something to stop students from blatant spam. If there's anyway to track how many messages X user has sent we can say that if they exceed a certain message threshold they get paused for a minute. Another thing is that I think messages really get lost when they start to stack up. There are 2 possible solutions either take the discord approach of doing different colors (probably won't work due to sheer number of users per live chat) or the second option is to implement 2 changes 1 is to highlight in some shade of yellow, and allow instructors to  pin/highlight comments of the instructors choice and give them some special shade as well.

This is an excellent addition to submitty and haas the potential to positively impact many users experience.

cjreed121 commented:
This is on hold of being merged till https://github.com/Submitty/Submitty/issues/10296 is closed

williamschen23 reviewed (changes requested):
all routes should be refactored to use php8 syntax
Example: `#[Route("/home/courses", methods: ["GET"])]`

zacharymnp reviewed (changes requested):
There is some unusual behavior that should be corrected.

A chatroom can be toggled between anonymous and not anonymous, but the checkbox is always unchecked. An anonymous chatroom needs to be checked on, then back off to turn off anonymous join.
https://github.com/user-attachments/assets/3d4512cd-40f5-48ea-a9f6-ea707b2866de

Leaving and then re-entering a chat changes any message from the user from USERNAME to 'me'.
https://github.com/user-attachments/assets/16ec65c6-4031-4988-bc19-d50f1b423cd2

Leaving and re-entering a chatroom as an anon will give a new ID every time. This may be intentional, but I think it should be consistent if it is possible.
https://github.com/user-attachments/assets/5488872b-57c6-42aa-b165-6592bae1cae4

jeffrey-cordero commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,267 @@
+            $chatroom->setDescription($_POST['description']);
+        }
+        if (isset($_POST['allow-anon'])) {
+            $chatroom->setAllowAnon(true);
```
We could simplify this into one line - `$chatroom->setAllowAnon(isset($_POST['allow-anon']));`

jeffrey-cordero commented on site/app/templates/chat/AllChatroomsPage.twig:
```diff
@@ -0,0 +1,119 @@
+                </thead>
+                <tbody>
+                    {% for chatroom in chatrooms %}
+                        {% if chatroom.isActive() %} <!--students only have access to active chat rooms-->
```
Would additional functionality to allow instructors to disable new messages, but still allow students to read prior messages, be a good addition to the current work?

jeffrey-cordero commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,237 @@
+// eslint-disable-next-line no-unused-vars
+function fetchMessages(chatroomId, my_id, when = new Date(0)) {
+    $.ajax({
+        // eslint-disable-next-line no-undef
```
You can remove this ESLint line if you add `buildCourseUrl` to the global line above. Same goes for  `displayErrorMessage` - `/* global csrfToken, buildCourseUrl, displayErrorMessage */`

jeffrey-cordero commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,237 @@
+function appendMessage(displayName, role, ts, content) {
+    let timestamp = ts;
+    if (!timestamp) {
+        timestamp = new Date(Date.now()).toLocaleString('en-us', { year: 'numeric', month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric' });
```
This block could be simplified with input of `ts || Date.now()` for the `Date` constructor. I don't see when `ts` would be `undefined` as it's based on well-formatted `GET` request responses or WebSocket messages from the server.

jeffrey-cordero commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,237 @@
+        success: function (response) {
+            try {
+                // eslint-disable-next-line no-unused-vars
+                const json = JSON.parse(response);
```
Wouldn't checking the `status` property from the server response be more reasonable here? It'd also align with the other methods in the file.

jeffrey-cordero commented on migration/migrator/migrations/course/20250325021641_lecture_chat.py:
```diff
@@ -0,0 +1,68 @@
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("DROP TABLE IF EXISTS chatrooms cascade")
```
Would manual dropping of the tables be more appropriate for keeping an existing log of these chatroom messages?

martig7 replied:
That would make more sense yeah.

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,108 @@
+.chatroom-container {
```
I believe it would be more user-friendly to make the container expand the entire width. A better divider between the messages and chatroom header would also be nice, if possible.

![image](https://github.com/user-attachments/assets/ef7f7e56-4bac-43ef-9a27-82f25e217e3e)

jeffrey-cordero reviewed (changes requested):
This is a solid implementation so far for adding live lecture chatrooms into Submitty. I've left many comments related to improvements that could be made, mainly related to the styling of the components or code optimizations that should be looked into.

jeffrey-cordero commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,264 @@
+        $message = new Message($msg_json['user_id'], $msg_json['display_name'], $msg_json['role'], $msg_json['content'], $chatroom);
+
+        try {
+            $client = new Client($this->core);
```
Forgot to mention this, but I would recommend having this in a separate function to handle setting up and sending WS messages with method signature `private function sendSocketMessage(array $msg_array): void`. I'd look into `OfficHoursQueueController.php` to see how to implement this to ensure consistency across the codebase.

jeffrey-cordero commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,264 @@
+
+        $hostId = $this->core->getUser()->getId();
+        $hostName = $this->core->getUser()->getDisplayFullName();
+        $chatroom = new Chatroom($hostId, $hostName, $_POST['title'], $_POST['description']);
```
Please ensure to return an error if `$_POST['title']` is an empty string after trimming it.

jeffrey-cordero commented on site/app/templates/chat/Chatroom.twig:
```diff
@@ -0,0 +1,26 @@
+        </div>
+        <div class="chatroom-content">
+            <div class="messages-area"></div>
+            <div class="input-container">
```
You can fix the responsiveness to appear as shown below. You'd override the `input-container` or add an additional class to have `min-width: none; flex: 1;`.

![image](https://github.com/user-attachments/assets/bac5d1ca-90c5-459d-82d2-7a81dbb01cff)

jeffrey-cordero commented on site/app/templates/chat/Chatroom.twig:
```diff
@@ -0,0 +1,26 @@
+    <div class="chatroom-container">
+        <div class="chatroom-header">
+            <span class="chatroom-title" title="{{ chatroom.getTitle() }}"> {{ chatroom.getTitle() | length > 40 ? chatroom.getTitle() | slice(0, 40) ~ '...' : chatroom.getTitle() }} </span>
+            <a href="{{ base_url }}" class="leave-room btn btn-danger">leave</a>
```
This should be capitalized.

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,118 @@
+}
+
+.message-header {
+    margin-bottom: 3px;
```
The header box shadow only seems to apply in dark mode and disappears when hovering over it. I'd recommend removing it entirely, as the divider is not necessarily needed, given vertical overflow scrolling exists.

![image](https://github.com/user-attachments/assets/68df8890-209f-4bb8-b627-fcfa5cc785f3)

martig7 replied:
I changed the header shadow to present in both light mode and dark mode. I like the header shadow, but I agree that it shouldn't disappear on hover

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,118 @@
+    overflow: hidden;
+    font-size: 1rem;
+    min-width: 85%;
+    margin-left: 1rem;
```
A `margin-right: 1rem` would ensure consistency with the input margin as well.

![image](https://github.com/user-attachments/assets/71b470d3-20fe-4762-93fe-0a822eb01535)

jeffrey-cordero commented on site/app/templates/chat/AllChatroomsPage.twig:
```diff
@@ -0,0 +1,119 @@
+                                {% else %}
+                                    <button class="btn btn-danger" onclick="toggle_chatroom({{ chatroom.getId() }}, true)">
+                                        <i class="fas fa-pause"></i>
+                                        End Session
```
I'd recommend sending out a WS message when a session ends or is deleted, which is similar to what `PollController.php` does. You'd disable inputs when the session ends or redirect users when a chatroom is deleted with a notice message.

martig7 replied:
After much suffering, it has been done.

martig7 commented on site/app/templates/chat/Chatroom.twig:
```diff
@@ -0,0 +1,26 @@
+        <div class="chatroom-content">
+            <div class="messages-area"></div>
+            <div class="input-container">
+                    <textarea name="content" class="message-input" placeholder="Enter your message here" maxlength=1500></textarea>
```
Do we want to remove the ability for users to enter long text? I guess there's a concern about spam but allowing new lines would allow for code to be typed in the chat which I think would be a common use case.

martig7 commented:
> @martig7 Are the screenshots in the PR description still up-to-date?

All but the chat screenshot, I'll update that now.

cjreed121 commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,284 @@
+    const tr = document.createElement('tr');
+    tr.id = `chatroom-row-${msg.id}`;
+
+    tr.innerHTML = `
+        <td>
+            <span class="display-short" title="${msg.title}">
+                ${msg.title.length > 30 ? `${msg.title.slice(0, 30)}...` : msg.title}
+            </span>
+        </td>
+        <td>
+            ${msg.host_name}
+        </td>
+        <td>
+            <span class="display-short" title="${msg.description}">
+                ${msg.description.length > 45 ? `${msg.description.slice(0, 45)}...` : msg.description}
+            </span>
+        </td>
+        <td>
+            <a href="${msg.base_url}/${msg.id}" class="btn btn-primary">Join</a>
+            ${msg.allow_anon ? `<i> or </i><a href="${msg.base_url}/${msg.id}/anonymous" class="btn btn-default">Join As Anon.</a>` : ''}
+        </td>
+    `;
```
I placed the do not merge label because of this comment. Please do not self-resolve this comment and verify with me that it is properly addressed.

There is an XSS security vuln here. Need to do this functionality without using `.innerHTML`. Some good options are to manually create the elements in JS and use that to inject in (can be a bit of a pain) or to use something like Vue.js. See calendar.js as an example for the first one.

cjreed121 replied:
Yes this should be safe. Thanks for addressing.

cjreed121 commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,303 @@
+}
+
+function appendMessage(displayName, role, ts, content) {
+    const timestamp = ts || new Date(Date.now()).toLocaleString('en-us', { year: 'numeric', month: 'short', day: 'numeric', hour: 'numeric', minute: 'numeric', second: 'numeric' });
```
Way too long of a line

cjreed121 commented on site/app/entities/UserEntity.php:
```diff
@@ -135,6 +135,7 @@ public function getDisplayInfo(): array {
         $out = [];
         $out["given_name"] = $this->user_preferred_givenname ?? $this->user_givenname;
         $out["family_name"] = $this->user_preferred_familyname ?? $this->user_familyname;
+        $out["full_name"] = $out["given_name"] . ' ' . $out["family_name"];
```
I would just recreate `getDisplayFullName()` from User.php and put it in this class.

martig7 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,284 @@
+        $display_name = '';
+        if ($chatroom->isAllowAnon() && isset($_SESSION[$boolKey]) && $_SESSION[$boolKey] === true) {
+            $display_name = $_SESSION[$sessKey];
+        }
```
I just made a lot of changes to address this, I think my approach is good but you should definitely check.

martig7 commented on site/app/entities/chat/Chatroom.php:
```diff
@@ -0,0 +1,114 @@
+    private string $host_id;
+
+    #[ORM\Column(type: Types::STRING)]
+    private string $host_name;
```
Alright, I've switched to eager fetching to prevent that second query.

cjreed121 commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,282 @@
+        $msg_array = [];
+        $msg_array['content'] = $_POST['content'];
+        $msg_array['type'] = 'chat_message';
+        $msg_array['user_id'] = $user->getId();
```
If you are going to send `user_id`, that removes all form of anonymity as a user could just open up chrome dev tools and see the user id.

cjreed121 commented on site/app/entities/chat/Chatroom.php:
```diff
@@ -0,0 +1,132 @@
+        $adjectives = ["Quick","Lazy","Cheerful","Pensive","Mysterious","Bright","Sly","Brave","Calm","Eager","Fierce","Gentle","Jolly","Kind","Lively","Nice","Proud","Quiet","Rapid","Swift"];
+        $nouns      = ["Duck","Goose","Swan","Eagle","Parrot","Owl","Sparrow","Robin","Pigeon","Falcon","Hawk","Flamingo","Pelican","Seagull","Cardinal","Canary","Finch","Hummingbird"];
+        $session_started_at = $this->getSessionStartedAt() !== null ? $this->getSessionStartedAt()->format('Y-m-d H:i:s') : 'unknown';
+        $seed_string = $user_id . '-' . $this->getId() . '-' . $this->getHostId() . '-' . $session_started_at;
+        $adj_hash = crc32($seed_string);
+        $noun_hash = crc32(strrev($seed_string));
+        $adj_index = abs($adj_hash) % count($adjectives);
+        $noun_index = abs($noun_hash) % count($nouns);
```
This is very easy to reverse engineer. All components to this are identifiable or very easily guessable. Meaning given a list of user_id's, an external user (knowing this code), could easily calculate everyone's anonymous name.

bmcutler reviewed (approved):
I tested this and I'm happy with the level of functionality.  We can add/modify features in later PRs.

In regards to the reverse engineering of the anonymous ids, we can improve that later.  It does have the start time in the hash so it isn't trivial to reverse engineer.  Related to that, when I stop and re-start the same chat room, because the timestamp changes, the anonymous ids will all change -- and that's weird.  So we probably want to modify it so at least for a single chat room the id stays the same.

jeffrey-cordero commented on site/app/entities/UserEntity.php:
```diff
@@ -120,6 +120,20 @@ public function getId(): string {
+    public function getDisplayedGivenName(): string {
+        return $this->user_preferred_givenname ?? $this->user_givenname;
+    }
+    public function getDisplayedFamilyName(): string {
```
A space between the methods here would be nice.

jeffrey-cordero commented on site/app/controllers/ChatroomController.php:
```diff
@@ -0,0 +1,282 @@
+            $msg_array['type'] = 'chat_close';
+            $msg_array['id'] = $chatroom->getId();
+            $msg_array['socket'] = "chatrooms";
+            $indiv_msg_array = [];
```
Add a comment here for clarity in socket pages.

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,113 @@
+    font-size: 1rem;
+    flex: 1;
+    margin-left: 1rem;
+    margin-right: 1rem;
```
I would update the `margin-right` to `0.5rem` and remove the `margin-left` from the `.send-message-btn` to reduce the space between them.

![image](https://github.com/user-attachments/assets/467dd4d8-b03b-4598-9260-7751b6bf692b)

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,114 @@
+.message-input {
+    border-radius: 5px;
+    outline: none;
+    resize: none;
```
I would remove the resize for now until Markdown is implemented in a future PR to handle large inputs.

jeffrey-cordero commented on site/public/js/chatroom.js:
```diff
@@ -0,0 +1,306 @@
+}
+
+function appendMessage(displayName, role, ts, content) {
+    const timestamp = ts || new Date(Date.now()).toLocaleString('en-us', 
```
The output of the right-hand side timestamp differs from the `ts` input. Over the network, it's in the form `2025-06-23 08:16:04 (ts)`, but when you call it with the `Date` constructor, it becomes a localized string, where for me it's in the form `'Jun 23, 2025, 8:17:07 AM'`. I'd suggest removing the right-hand side here because `ts` should never be undefined, as it comes from the server.

jeffrey-cordero commented on site/public/css/chatroom.css:
```diff
@@ -0,0 +1,114 @@
+}
+
+.send-message-btn {
+    margin-left: 1rem;
```
I'd suggest removing the left margin to remove the unnecessary gap.

![image](https://github.com/user-attachments/assets/0cef8124-a19c-4dde-8abe-815425f8190c)

![image](https://github.com/user-attachments/assets/4e9520d6-db59-4df9-9b51-4f13664d0b4a)
